ddht/v5_1/alexandria/resource_queue.py

```diff
@@ -0,0 +1,56 @@
+from typing import Any, AsyncIterator, Collection, Container, Hashable, Set, TypeVar
+
+from async_generator import asynccontextmanager
+import trio
+
+TResource = TypeVar("TResource", bound=Hashable)
+
+
+class RemoveResource(Exception):
+    pass
+
+
+class ResourceQueue(Container[TResource]):
+    resources: Set[TResource]
+
+    def __init__(
+        self, resources: Collection[TResource], max_resource_count: int = 256
+    ) -> None:
+        if len(resources) > max_resource_count:
+            raise ValueError(
+                f"Number of resources exceeds maximum: {len(resources)} > {max_resource_count}"
+            )
+        self.resources = set(resources)
+        self._max_resource_count = max_resource_count
+        self._send, self._receive = trio.open_memory_channel[TResource](
+            max_resource_count
+        )
+        for resource in resources:
+            self._send.send_nowait(resource)
+
+    async def add(self, resource: TResource) -> None:
+        if resource in self:
+            raise ValueError("Resource already tracked")
+        self.resources.add(resource)
+        await self._send.send(resource)
+
+    def __contains__(self, value: Any) -> bool:
+        return value in self.resources
+
+    def remove(self, resource: TResource) -> None:
+        self.resources.remove(resource)
+
+    @asynccontextmanager
+    async def reserve(self) -> AsyncIterator[TResource]:
+        while True:
+            resource = await self._receive.receive()
+            if resource in self:
+                break
+            else:
+                continue
```

Comment:
When will the resource not be `in self`?

Reply:
If `queue.remove()` is called within the context block.

```python
async with queue.reserve() as resource:
    ...  # do something with it
    if some_condition:
        queue.remove(resource)
```

Reply:
Ahh, good point